### PR TITLE
Allow Type Reference in the Redis data source

### LINKED ISSUE #29865: Better support for Generic with redis#pubsub group
### Description

Today, the pubsub group of redis uses a Class<V>, which is not ideal for use with generics

### Implementation ideas

Use jackson TypeReference as the redis extension already depends on it

### LINKED ISSUE #34554: ReactiveHashCommands doesn't properly encode/decode Object value
### Describe the bug

`ReactiveHashCommands<String, String, Object>` uses [Marshaller](https://github.com/quarkusio/quarkus/blob/3.1.3.Final/extensions/redis-client/runtime/src/main/java/io/quarkus/redis/runtime/datasource/Marshaller.java),  this class encodes String into plain string without quotes, not JSON string,  but decodes with Object.class.

I have to use `ReactiveHashCommands<String, String, Object[]>` to bypass,  any chance to make `ReactiveHashCommands<String, String, Object>` just works?

### Quarkus version or git rev

3.1

### PR DESCRIPTION
**Pull request by cescoffier**
Allows configuring the Redis data source with type references

The Redis data source used Class<?> to configure the type safety. Unfortunately, it didn't work with generics. 

This commit adds methods using (Jackson) TypeReference to configure the various groups. Thus, generics are now supported.

Also:

- deprecate the graph group, as Redis declared it is as end-of-life ( :-( )
- deprecate search(Class<K> keyType) as the index must be a string
- re-structure a bit the Redis reference guide 

Fix https://github.com/quarkusio/quarkus/issues/34554
Fix https://github.com/quarkusio/quarkus/issues/29865

### CODE REVIEW

File: docs/src/main/asciidoc/redis-reference.adoc [hunk 1]
```diff
@@ -370,7 +377,37 @@ The `canHandle` method is called to check if the codec can handle a specific typ
...
+        commands.hset(MY_KEY, field, value);
+    }
+
+    public List<Person> get(String field) {
+        return commands.hget(MY_KEY, field);
+    }
+}
+----
+
+IMPORTANT: You cannot use type references when using transaction. This is a known limitation.
```
Comment (machi1990) [inline docs/src/main/asciidoc/redis-reference.adoc:408]: For my own knowledge, where is the limitation coming from?

File: docs/src/main/asciidoc/redis-reference.adoc [hunk 2]
```diff
@@ -205,14 +205,21 @@ quarkus.redis.hosts=<the redis cloud url such as redis://redis-12436.c14.us-east
...
 ----
 
-=== Authentication
+=== Use TLS
+
+To use TLS, you need:
+
+1. Set the `quarkus.redis.tls.enabled=true` property
+2. Make sure that your URL starts with `rediss://` (with two `s`)
+
```
Comment (machi1990) [inline docs/src/main/asciidoc/redis-reference.adoc:214]: Nice. This takes care of this suggestion https://github.com/quarkusio/quarkus/issues/34761#issuecomment-1637126882 as well. Thanks

**Review by machi1990 — APPROVED** [review]

File: docs/src/main/asciidoc/redis-reference.adoc (hunk 1 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/redis-reference.adoc:408]: Time :-D 

It is technically possible to add support for type reference when using transactions. However, we never had that specific request.

File: docs/src/main/asciidoc/redis-reference.adoc (hunk 2 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/redis-reference.adoc:214]: Exactly, I was on it, so I added it.

File: docs/src/main/asciidoc/redis-reference.adoc (hunk 1 above)
Comment (machi1990) [reply docs/src/main/asciidoc/redis-reference.adoc:408]: I thought so as well, thanks for the reply. Makes sense, and when the request comes, we can build on top of this PR.

**Pull request merged**